cirq/contrib/quirk/cells/parse.py

```diff
@@ -12,121 +12,376 @@
     except Exception as ex:
-        raise ValueError(
-            'Failed to parse complex from {!r}'.format(text)) from ex
-
-
-def _parse_scalar(text: str, **extras) -> sympy.Basic:
-    text = _expand_unicode_fractions(text)
-
-    # Convert number suffixing into implicit action.
-    text = re.sub(r'([0123456789)])([a-df-zA-DF-Z])', r'\1 \2', text)
-
-    # Canonicalize redundant function names.
-    text = text.replace('√', ' sqrt')
-    text = text.replace('arcsin', 'asin')
-    text = text.replace('arccos', 'acos')
-    text = text.replace('arctan', 'atan')
-
-    # Convert implicit function application into explicit application.
-    # TODO(craiggidney): support nested implicit function application.
-    text = re.sub(r'(a?(cos|sin|tan)|exp|ln|sqrt)\s*([^-\s(+*/^)]+)', r'\1(\3)',
-                  text)
-
-    # Convert implicit multiplication into explicit multiplication.
-    text = re.sub(r'([^-^+*/])\s([^-^+*/])', r'\1*\2', text)
-
-    # Translate exponentiation notation.
-    text = text.replace('^', '**')
-
-    return parse_expr(text,
-                      transformations=[],
-                      local_dict={},
-                      global_dict={
-                          'sqrt': sympy.sqrt,
-                          'exp': sympy.exp,
-                          'ln': sympy.ln,
-                          'cos': sympy.cos,
-                          'sin': sympy.sin,
-                          'tan': sympy.tan,
-                          'acos': sympy.acos,
-                          'asin': sympy.asin,
-                          'atan': sympy.atan,
-                          'i': sympy.I,
-                          'e': sympy.E,
-                          'pi': sympy.pi,
-                          **extras,
-                      })
-
-
-def _expand_unicode_fractions(text: str) -> str:
-    for k, v in UNICODE_FRACTIONS.items():
-        text = text.replace(k, v)
-    return text
+        raise ValueError(f'Failed to parse complex from {text!r}') from ex
 
 
-UNICODE_FRACTIONS = {
-    "½": "(1/2)",
-    "¼": "(1/4)",
-    "¾": "(3/4)",
-    "⅓": "(1/3)",
-    "⅔": "(2/3)",
-    "⅕": "(1/5)",
-    "⅖": "(2/5)",
-    "⅗": "(3/5)",
-    "⅘": "(4/5)",
-    "⅙": "(1/6)",
-    "⅚": "(5/6)",
-    "⅐": "(1/7)",
-    "⅛": "(1/8)",
-    "⅜": "(3/8)",
-    "⅝": "(5/8)",
-    "⅞": "(7/8)",
-    "⅑": "(1/9)",
-    "⅒": "(1/10)",
-}
+def parse_formula(formula: str) -> Union[float, sympy.Basic]:
+    """Attempts to parse formula text in exactly the same way as Quirk."""
+    if not isinstance(formula, str):
+        raise TypeError('formula must be a string')
+
+    token_map = {**PARSE_COMPLEX_TOKEN_MAP_RAD, 't': sympy.Symbol('t')}
+    result = _parse_formula_using_token_map(formula, token_map)
+
+    if isinstance(result, sympy.Basic):
+        if result.free_symbols:
+            return result
+        result = complex(result)
+
+    if isinstance(result, complex):
+        if abs(np.imag(result)) > 1e-8:
```

Comment:
nit: pass in tolerance for consistency with rest of cirq

Reply:
Acknowledged, but that would just be a "be inconsistent with Quirk" parameter.